File: Data Structures & Algorithms/binary-search/submission-1.java
class Solution {
    public int search(int[] nums, int target) {
        return bs(nums, 0, nums.length, target);
    }

    public int bs(int[] nums, int start, int stop, int target) {
        if (start >= stop) {
            return -1;
        }

        int mid = (stop - start) / 2 + start;
        if (target == nums[mid]) {
            return mid;
        } else if (target < nums[mid]) {
            return bs(nums, 0, mid, target);
        } else {
            return bs(nums, mid + 1, stop, target);
        }
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        sol = set()
        nums.sort()

        for mid in range(1, len(nums) - 1):
            left = 0
            right = len(nums) - 1
            while left < right and left < mid and mid < right:
                if nums[mid] + nums[left] + nums[right] == 0:
                    sol.add((nums[left], nums[mid], nums[right]))
                    if left + 1 == mid:
                        left = mid + 1
                    else: 
                        left += 1
                    if right - 1 == mid:
                        right = mid - 1
                    else: 
                        right -= 1
                elif nums[mid] + nums[left] + nums[right] < 0:
                    if left + 1 == mid:
                        left = mid + 1
                    else:
                        left += 1
                else:
                    if right - 1 == mid:
                        right = mid - 1
                    else: 
                        right -= 1
        return [list(tup) for tup in sol]